File: backend/src/main/java/com/clinica/backend/security/Roles.java
package com.clinica.backend.security;

import java.util.Set;

/**
 * Roles del sistema. Son combinables: por ejemplo, un psicólogo que además administra la
 * clínica tiene {@link #PROFESIONAL} y {@link #ADMIN}. La especialidad (Psicología,
 * Dermatología) es un atributo del usuario, no un rol.
 * <ul>
 *   <li>{@link #PROFESIONAL}: profesional de salud; único rol con acceso a la información
 *   clínica y el que aparece en la agenda.</li>
 *   <li>{@link #ADMIN}: administra la clínica de su especialidad (usuarios, cobros,
 *   inventario, catálogos, eliminación de pacientes y cobros). Solo con {@link #PROFESIONAL}
 *   además puede ver o gestionar registros clínicos de otros profesionales.</li>
 *   <li>{@link #ASISTENTE}: recepción; pacientes, citas y cobros, sin información clínica.</li>
 *   <li>{@link #SITE_ADMIN}: administración del sitio web público; sin acceso clínico.</li>
 * </ul>
 */
public final class Roles {

    public static final String ADMIN = "ROLE_ADMIN";
    public static final String PROFESIONAL = "ROLE_PROFESIONAL";
    public static final String ASISTENTE = "ROLE_ASISTENTE";
    public static final String SITE_ADMIN = "ROLE_SITE_ADMIN";

    public static final Set<String> ALL = Set.of(ADMIN, PROFESIONAL, ASISTENTE, SITE_ADMIN);

    private Roles() {
    }
}

File: frontend/src/app/core/models/roles.ts
/**
 * Roles del sistema (espejo de `Roles.java`). Son combinables: por ejemplo, una psicóloga que
 * además administra la clínica tiene PROFESIONAL y ADMIN. La especialidad es un atributo del
 * usuario, no un rol.
 */
export const ROLES = {
  PROFESIONAL: 'ROLE_PROFESIONAL',
  ADMIN: 'ROLE_ADMIN',
  ASISTENTE: 'ROLE_ASISTENTE',
  SITE_ADMIN: 'ROLE_SITE_ADMIN'
} as const;

export interface RoleOption {
  code: string;
  label: string;
  description: string;
  badgeClass: string;
}

/** Roles que un administrador de la clínica puede asignar desde la gestión de usuarios. */
export const ASSIGNABLE_ROLES: RoleOption[] = [
  {
    code: ROLES.PROFESIONAL,
    label: 'Profesional de salud',
    description: 'Atiende pacientes y accede a la historia clínica de su especialidad.',
    badgeClass: 'bg-primary-50 text-primary-700 border-primary-200'
  },
  {
    code: ROLES.ADMIN,
    label: 'Administrador',
    description: 'Gestiona usuarios, cobros, inventario y catálogos; puede eliminar pacientes y cobros.',
    badgeClass: 'bg-violet-50 text-violet-700 border-violet-200'
  },
  {
    code: ROLES.ASISTENTE,
    label: 'Asistente / Recepción',
    description: 'Gestiona pacientes, citas y cobros, sin acceso a información clínica.',
    badgeClass: 'bg-slate-100 text-slate-700 border-slate-200'
  }
];

const SITE_ADMIN_OPTION: RoleOption = {
  code: ROLES.SITE_ADMIN,
  label: 'Administrador web',
  description: 'Administra el sitio web público.',
  badgeClass: 'bg-amber-50 text-amber-700 border-amber-200'
};

/** Roles del usuario en orden estable, para mostrarlos como etiquetas. */
export function describeRoles(roles: readonly string[] | undefined): RoleOption[] {
  return [...ASSIGNABLE_ROLES, SITE_ADMIN_OPTION].filter(option => roles?.includes(option.code));
}

/** Misma regla que el backend: al menos un rol y Profesional no se combina con Asistente. */
export function roleSelectionError(roles: readonly string[] | undefined): string | null {
  if (!roles || roles.length === 0) {
    return 'Seleccione al menos un rol.';
  }
  if (roles.includes(ROLES.PROFESIONAL) && roles.includes(ROLES.ASISTENTE)) {
    return 'Un profesional ya puede gestionar pacientes, citas y cobros: no combine Profesional con Asistente.';
  }
  return null;
}
